metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/DereferenceUtils.java

```diff
@@ -1,191 +1,204 @@
+     *
+     * @param rdf input document
+     * @return non-null set of values for dereferencing, not containing null.
+     */
+    public static Map<Class<? extends AboutType>,Set<String>> extractReferencesForDereferencing(RDF rdf) {
+
+        // Get all the links we are interested in.
+        final Map<Class<? extends AboutType>,Set<String>> result = new HashMap<>();
+        extractValues(rdf.getAgentList(), item -> dereferenceAgent(item, result));
+        extractValues(rdf.getConceptList(), item -> dereferenceConcept(item, result));
+        extractValues(rdf.getPlaceList(), item -> dereferencePlace(item, result));
+        extractValues(rdf.getTimeSpanList(), item -> dereferenceTimespan(item, result));
+        extractValues(rdf.getWebResourceList(), item -> dereferenceWebResource(item, result));
+        extractValues(rdf.getAggregationList(), item -> dereferenceAggregation(item, result));
+        extractValues(RdfEntityUtils.getProviderProxies(rdf), item -> dereferenceProxy(item, result));
+
+        // Clean up the result: no null values and no objects that we already have.
+        result.remove(null);
+        final Consumer<List<? extends AboutType>> cleaner = list -> Optional.ofNullable(list)
+                .map(List::stream).orElseGet(Stream::empty).map(AboutType::getAbout)
+                .forEach(result::remove);
+        cleaner.accept(rdf.getAgentList());
+        cleaner.accept(rdf.getConceptList());
+        cleaner.accept(rdf.getPlaceList());
+        cleaner.accept(rdf.getTimeSpanList());
+
+        // Done.
+        return result;
+    }
+
+    private static <S> void extractValues(List<S> source, Consumer<S> extractor) {
+        if (source != null) {
+            for (S sourceItem : source) {
+                extractor.accept(sourceItem);
+            }
+        }
+    }
+
+    private static <T> void convertValues(List<? extends T> source,
+                                          Function<T, String> conversion, Map<Class<? extends AboutType>,Set<String>> result,
+                                          Class<? extends AboutType> classType) {
+        if (source != null) {
+            for (T sourceItem : source) {
+                convertValue(sourceItem, conversion, result, classType);
+            }
+        }
+    }
+
+    private static <T> void convertValue(T source, Function<T, String> conversion,
+                                         Map<Class<? extends AboutType>,Set<String>> resultMap,
+                                         Class<? extends AboutType> classType) {
+        Set<String> result = new HashSet<>();
+        if (source != null) {
+            final String target = conversion.apply(source);
+            if (StringUtils.isNotEmpty(target)) {
+                result.add(target);
+                //If Map already contains key with 'classType', then just add the new values into existing Set
+                if(resultMap.containsKey(classType)) {
```

Comment:
I've added it! I also added `resultMap.putIfAbsent(classType, result)` in case it is not in the map.
From what I read about ` computeIfPresent`, it does not add the elem if not present

Reply:
Yes that is one way to go!